text/0000-ember-data-model-factory-for.md

```diff
@@ -0,0 +1,189 @@
+
+Users wishing to extend the behavior of `modelFactoryFor` could do so in the following manner:
+
+**Example 1:**
+
+**services/store.js**
+```js
+import { getOwner } from '@ember/application';
+import Store from 'ember-data/store';
+
+export default Store.extend({
+  modelFactoryFor(modelName) {
+    if (someCustomCondition) {
+      return getOwner(this).factoryFor(someFactoryName);
+    }
+    
+    return this._super(modelName);
+  }
+});
+```
+ 
+ #### `Model.modelName`
+ 
+ `ember-data` currently sets `modelName` onto the `klass` accessible via the `factory`. For classes that do not
+   inherit from `DS.Model` this would not be done, although end users may do so themselves in their implementations
+   if so desired.
+
+### What is a valid factory?
+
+The default export of a custom ModelClass **MUST** conform to the requirements of `Ember.factoryFor`. The requirements
+ of `factoryFor` are currently underspecified; however, in practice, this means that the default export is an
+ instantiable class with a static `create` method and an instance `destroy` method or that inherits from `EmberObject`
+ (which provides such methods).
+ 
+**Example 2:**
+
+```javascript
+import { assign } from '@ember/polyfills';
+
+export default class CustomModel {
+  constructor(createArgs) {
+    assign(this, createArgs);
+  }
+  destroy() {
+    // ... do teardown
+  }
+  static create(createArgs) {
+    return new this(createArgs);
+  }
+}
+```
+
+**Example 3:**
+
+```javascript
+import EmberObject from '@ember/object';
+
+export default class CustomModel extends EmberObject {
+  constructor(createArgs) {
+    super(createArgs);
+  }
+}
+```
+
+Custom classes for models should expect their constructor to receive a single argument: an object with *at least*
+ the following.
+ 
+- A `recordData` instance available on the `recordData` `Symbol` and accessible via `getRecordData` (see below)
+- Any properties passed as the second arg to `createRecord`
+- An `owner` available on the owner `Symbol` and accessible via `Ember.getOwner`
+- Any DI injections
+- any other properties that `Ember` chooses to pass to a class instantiated via `factory.create` (currently none)
+
+### getRecordData/setRecordData
+
+Every `record` (instance of the class returned by `modelFactoryFor`) will have an associated [RecordData](https://github.com/emberjs/rfcs/pull/293)
+ which contains the backing data for the id, type, attributes and relationships of that record.
+ 
+This backing data can be accessed by using the `getRecordData` util on the `record` (or on the `createArgs` passed to
+ a record).
+ 
+If providing a custom class that does not inherit from `EmberObject`, you *MUST* either use `assign` (see example 2)
+ or `setRecordData` (see example 4) to add the symbol for the backing `recordData` instance onto your record instance.
+ If inheriting from `EmberObject` calling `super` with the `createArgs` is sufficient.
```

Comment:
Why is it necessary for the record to have a backref to the record data?

We should certainly recommend that people do this, but I don't see why we should require it.  I'd also recommend we make the public api here `setRecordData` and not allow the record data to be set via `assign`, to simply the docs if nothing else.

Reply:
@hjdivad implementations of `Model` and `RecordData` need to be able to convert from a `record` to a `record-data` when `createRecord` is called or when a relationship is mutated. 

Overtime we may be able to loosen that requirement, although personally I think that even if no longer necessary it is a good one.

RE `assign`, since we are using a `Symbol`, `Object.assign` and the Ember polyfill `"just work"`. `Object.keys` and `for ... in` do not work unless using Ember's `Symbol` polyfill.

<img width="370" alt="screen shot 2018-10-04 at 1 42 44 pm" src="https://user-images.githubusercontent.com/650309/46501835-746f0c00-c7db-11e8-8a12-cddb4e92cc71.png">

Reply:
Yes, `assign` can work but it will be simpler if we have a single public API, `setRecordData`.  `assign` works as a side effect of using a `Symbol`, but this is an implementation detail (you could imagine supporting the APIs with `WeakMap`s for instance).

> @hjdivad implementations of Model and RecordData need to be able to convert from a record to a record-data when createRecord is called or when a relationship is mutated.

That's true for `DS.Model`s but need not be the case for other model classes.  That `record.set('belongsToRelationship', otherRecord)` invokes `recordData.setBelongsTo` is an implementation detail of `DS.Model` -- what part of the system necessitates that this is done by *all* model implementations?

I think it's important that we avoid these kinds of requirements to allow users an escape valve for relationship semantics when they need it.

Reply:
> I think it's important that we avoid these kinds of requirements to allow users an escape valve for relationship semantics when they need it.

I feel as though we won't be able to have any-level of decent interop between various `RecordData` and `Model` implementations if we don't have this constraint. While I agree the existing relationship layer could (and should) be refactored to be less coupled, I think it's mechanics derives in part from the need for an interop story.

Reply:
> Yes, assign can work but it will be simpler if we have a single public API, setRecordData. assign works as a side effect of using a Symbol, but this is an implementation detail (you could imagine supporting the APIs with WeakMaps for instance).

I can get behind this idea and will update the language.

Reply:
I think if we consider `recordData` as a "weak-mapped" thing (even if symbol based), we may be able to remove the requirement. Users would need to call `recordDataFor` on `createArgs` from within the constructor, but afterwards we could weakMap to the instance or symbolicate it ourselves and calls to `recordDataFor` on the instance would "just work"

Reply:
> I feel as though we won't be able to have any-level of decent interop between various RecordData and Model implementations if we don't have this constraint. While I agree the existing relationship layer could (and should) be refactored to be less coupled, I think it's mechanics derives in part from the need for an interop story.

That interop should be driven by `RecordData`s; the relationships already deal with these objects and should be as indifferent as can be to the particular records associated with some particular `RecordData`.

There's no dispute that `RecordData`s need to know what record to return at various times: the basic mechanism of caching an invocation of `factory.create(...)` works well here.  Model implementations can set up edges between `RecordData`s via the public `RecordData` API.  Why is that insufficient to achieve interop?

Reply:
@hjdivad because if you are a `record` and someone calls `record.set` with a different `record` then either the `record` or the `record-data` for the receiving `record` needs to be able to lookup the `record-data` instance associated with it. In absence of this ability, you end up needing to expose various other public interfaces to make things work. That said, I think the "weak-map after create" approach will allow us to make this work just fine.

Reply:
> @hjdivad because if you are a record and someone calls record.set with a different record then either the record or the record-data for the receiving record needs to be able to lookup the record-data instance associated with it.

Sure, so it's recommended to `setRecordData` to achieve this.  But that's not the same thing as it being required.  In particular I don't think the rest of the system should assume that records (besides `DS.Model`s) have a backref of any kind to their `RecordData`.